### PR TITLE
Preparatory work for Styled Components 5 & latest TS

### PR DESCRIPTION
**Pull request by lukelooker**
### :sparkles: Changes

This pulls in a small collection of changes that will be needed when we make the upgrade to Styled Components 5. 

While that upgrade is waiting for our repository migration I figured it'd make sense to incorporate these changes so we're ready to go when we can do the upgrade freely.

### :white_check_mark: Requirements

- [ ] Includes test coverage for all changes
- [x] Build and tests are passing
- [ ] Update documentation
- [ ] Updated CHANGELOG
- [ ] Checked for i18n impacts
- [ ] Checked for a11y impacts
- [x] PR is ideally < ~400LOC

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -73,6 +73,7 @@
     "prettier": "^2.0.5",
     "react": "^16.13.1",
     "react-dom": "^16.13.1",
+    "react-is": "^16.13.1",
```
Comment (lukelooker) [inline package.json:76]: SC5 requires this as dependency.

File: packages/components-providers/package.json [hunk 2]
```diff
@@ -22,10 +22,12 @@
     "@types/react": "^16.9.49",
     "@types/styled-components": "^4.4.1",
     "react": "^16.13.1",
+    "react-is": "^16.13.1",
     "styled-components": "^4.4.1"
   },
   "peerDependencies": {
-    "react": "^16.9.0",
+    "react": "^16.11",
```
Comment (lukelooker) [inline packages/components-providers/package.json:29]: Bring our peer dependency up to the same version used in the CORE product currently.

File: packages/components/src/Dialog/Surface.tsx [hunk 3]
```diff
@@ -52,7 +52,6 @@ interface SurfaceProps
     ColorProps,
     LayoutProps {
   surfaceStyles?: CSSObject
-  anchor?: 'right'
```
Comment (lukelooker) [inline packages/components/src/Dialog/Surface.tsx:55]: New TS checked found that this was unused.

File: packages/components/src/Form/Inputs/InputColor/InputColor.tsx [hunk 4]
```diff
@@ -135,7 +135,7 @@ export const InputColorComponent = forwardRef(
     }
 
     const handleColorChange = (hs: HueSaturation) => {
-      setColorState({ h: 0, s: 100, ...hs, v: get(color, 'v', 1) })
+      setColorState({ ...hs, v: get(color, 'v', 1) })
```
Comment (lukelooker) [inline packages/components/src/Form/Inputs/InputColor/InputColor.tsx:138]: New TS checked detects that the spread does the same thing as the `h: 0, s: 100` here.

File: packages/components/src/Link/Link.test.tsx [hunk 5]
```diff
@@ -43,7 +43,7 @@ describe('Link', () => {
   test('Underline', () => {
     renderWithTheme(<Link underline>My link</Link>)
     const link = screen.getByText('My link')
-    expect(link).toHaveStyle('text-decoration: underline')
+    expect(link).toHaveStyleRule('text-decoration: underline')
```
Comment (lukelooker) [inline packages/components/src/Link/Link.test.tsx:46]: Previous version doesn't actually verify (quirk)

File: packages/components/src/List/List.test.tsx [hunk 6]
```diff
@@ -68,15 +69,16 @@ test('A numerically ordered List', () => {
...
       <ListItem>🥑</ListItem>
       <ListItem>🍕</ListItem>
       <ListItem>🥨</ListItem>
     </List>
   )
-  const tree = component.toJSON()
-  expect(tree).toHaveStyleRule('list-style-type', 'none')
+
+  const list = screen.getByTestId('list')
+  expect(list).toHaveStyle('list-style-type: none')
```
Comment (lukelooker) [inline packages/components/src/List/List.test.tsx:81]: This test wasn't working before. Gave a false positive.

File: packages/components/src/utils/getCurrentNode.ts [hunk 7]
```diff
@@ -36,6 +36,7 @@ export function getCurrentNode<E extends HTMLElement = HTMLElement>(
   elementOrRefObject: ElementOrRef<E>
 ): E | null {
   if (!elementOrRefObject) return null
+  // @ts-expect-error: Need to find a better way to do union below
```
Comment (lukelooker) [inline packages/components/src/utils/getCurrentNode.ts:39]: TS 3.9 / 4.0 doesn't like this.

File: storybook/src/Forms/TextArea.stories.tsx [hunk 8]
```diff
@@ -33,7 +33,7 @@ export const All = () => (
     <Required />
     <Error />
     <Inline />
-    <Resize />>
+    <Resize />
```
Comment (lukelooker) [inline storybook/src/Forms/TextArea.stories.tsx:36]: Type. TS 3.9 catches this.

File: www/src/documentation/components/content/demos/CalendarPropTable.tsx [hunk 9]
```diff
@@ -55,7 +55,7 @@ export const CalendarPropTable = () => {
           <TableRow>
             <TableDataCell pr="large">onDayClick</TableDataCell>
             <TableDataCell pr="large">
-              <Code>(date: Date) => void</Code>
+              <Code>(date: Date) ={'>'} void</Code>
```
Comment (lukelooker) [inline www/src/documentation/components/content/demos/CalendarPropTable.tsx:58]: TS 3.9 does not like `>` symbol mid-JSX.

File: www/src/documentation/components/content/demos/CalendarPropTable.tsx [hunk 10]
```diff
@@ -133,10 +133,10 @@ export const CalendarPropTable = () => {
...
 
 const TableKey = styled.div`
-  border-top: 1px solid ${({ theme }) => theme.colors.ui2};
   background: ${({ theme }) => theme.colors.ui1};
+  border-top: 1px solid ${({ theme }) => theme.colors.ui2};
   color: ${({ theme }) => theme.colors.text3};
   font-weight: bold;
-  padding: 0.5rem;
   margin-bottom: 2rem;
+  padding: 0.5rem;
```
Comment (lukelooker) [inline www/src/documentation/components/content/demos/CalendarPropTable.tsx:141]: Stylelint demands alphabetical :)

**Review by dbchristopher — APPROVED** [review]
LGTM!

**Pull request merged**